### PR TITLE
Use context aware dial functions

### PR DESCRIPTION
**Pull request by mmatczuk**

### CODE REVIEW

File: dialvia/http.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"net/url"
+)
+
+type httpProxyDialer struct {
+	dial      ContextDialerFunc
+	proxyURL  *url.URL
+	tlsConfig *tls.Config
+}
+
+func HTTPProxy(dial ContextDialerFunc, proxyURL *url.URL) *httpProxyDialer {
```
Comment (Choraden) [inline dialvia/http.go]: Exported function returns unexported type.

File: proxy.go [hunk 2]
```diff
@@ -790,45 +790,23 @@ func (p *Proxy) connect(req *http.Request) (*http.Response, net.Conn, error) {
...
-	}
-	if proxyURL.User != nil {
-		connReq.Header.Add("Proxy-Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(proxyURL.User.String())))
-	}
-	connReq.Write(pbw)
-
-	if err := pbw.Flush(); err != nil {
-		conn.Close()
-		return nil, nil, err
+		d := dialvia.HTTPSProxy(p.dial, proxyURL, p.clientTLCConfig())
```
Comment (Choraden) [inline proxy.go]: Is `clientTLCConfig()` an intentional name?

**Review by Choraden — COMMENTED** [review]
Overall LGTM

File: dialvia/http.go (hunk 1 above)
Comment (mmatczuk) [reply dialvia/http.go]: Any alternatives?

File: proxy.go (hunk 2 above)
Comment (mmatczuk) [reply proxy.go]: Old typo, will fix it.

File: dialvia/http.go (hunk 1 above)
Comment (Choraden) [reply dialvia/http.go]: How about it returns Dialer interface?

Comment (mmatczuk) [reply dialvia/http.go]: We could pass the original response in a custom error type.

Comment (Choraden) [reply dialvia/http.go]: I would rather export it than wrap with custom error.

Comment (mmatczuk) [reply dialvia/http.go]: Exporting it gives no benefit except for satisfying linter, and pollutes the pkg api.

**Review by Choraden — APPROVED** [review]

**Pull request merged**